proxy/src/outbound.rs

```diff
@@ -51,81 +51,78 @@ impl<B> Outbound<B> {
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum Destination {
-    LocalSvc(FullyQualifiedAuthority),
-    External(SocketAddr),
-}
-
-impl From<FullyQualifiedAuthority> for Destination {
-    #[inline]
-    fn from(authority: FullyQualifiedAuthority) -> Self {
-        Destination::LocalSvc(authority)
-    }
+pub enum Dst {
+    Unbound(FullyQualifiedAuthority, Protocol),
+    Bound(SocketAddr, Protocol),
 }
 
-impl From<SocketAddr> for Destination {
-    #[inline]
-    fn from(addr: SocketAddr) -> Self {
-        Destination::External(addr)
-    }
-}
-
-
 impl<B> Recognize for Outbound<B>
 where
     B: tower_h2::Body + 'static,
 {
     type Request = http::Request<B>;
     type Response = bind::HttpResponse;
     type Error = <Self::Service as tower::Service>::Error;
-    type Key = (Destination, Protocol);
+    type Key = Dst;
     type RouteError = bind::BufferSpawnError;
     type Service = InFlightLimit<Timeout<Buffer<Balance<
         load::WithPendingRequests<Discovery<B>>,
         choose::PowerOfTwoChoices<rand::ThreadRng>
     >>>>;
 
+    /// Determines the destination service for a given Request.
+    ///
+    /// Destinations are either bound or unbound. Unbound destinations must be resolved
+    /// through the controller to produce a list of endpoints. Bound destinations use the
+    /// client-provided destination address.
+    ///
+    /// For HTTP/2 requests, a service may support several :authority values. HTTP/1
+    /// requests, on the other hand, may not share a service across authorities.
+    ///
+    /// Services may not be reused if they do not include an authority.
     fn recognize(&self, req: &Self::Request) -> Option<Reuse<Self::Key>> {
         let proto = bind::Protocol::detect(req);
 
-        // The request URI and Host: header have not yet been normalized
-        // by `NormalizeUri`, as we need to know whether the request will
-        // be routed by Host/authority or by SO_ORIGINAL_DST, in order to
-        // determine whether the service is reusable.
-        let local = req.uri().authority_part()
+        // Determine the request's `:authority` (or `Host`), either from its headers or
+        // from the request's URI.
+        //
+        // The authority is normalized to determine whether or not it should be resolved
+        // through the controller.
+        let qualified_name = req.uri().authority_part()
             .cloned()
-        // Therefore, we need to check the host header as well as the URI
-        // for a valid authority, before we fall back to SO_ORIGINAL_DST.
             .or_else(|| h1::authority_from_host(req))
             .map(|authority| {
-                FullyQualifiedAuthority::normalize(
-                    &authority,
-                    &self.default_namespace)
+                FullyQualifiedAuthority::normalize(&authority, &self.default_namespace)
```

Comment:
`h1::authority_from_hist(req)` does the `SO_ORIGINAL_DST` logic itself, so if it returns None then we can return None right away. That is, we just did the "If there is no original dst, then we have nothing!" part from below, and we don't need to do SO_ORIGINAL_DST stuff below, right?

Reply:
I don't think that's right.

`h1::normalize_our_view_of_uri` does `SO_ORIGINAL_DST` logic:
https://github.com/runconduit/conduit/blob/0fbcf956321b5ef1295319fa1d95d4694777038b/proxy/src/transparency/h1.rs#L20-L36

`h1::authority_from_host` does not:
https://github.com/runconduit/conduit/blob/0fbcf956321b5ef1295319fa1d95d4694777038b/proxy/src/transparency/h1.rs#L47-L60